### PR TITLE
[aptos-telemetry] telemetry service client integration

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR implements a sender to send telemetry to the telemetry web service in an authenticated way.

### Test Plan

Run in one terminal `cargo run -p aptos-telemetry-service -- -f telemetry-service.yaml`

Run in another terminal: `cargo run -p aptos-node -- --test`

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2763)

### CODE REVIEW

File: crates/aptos-telemetry/src/metrics.rs [hunk 1]
```diff
@@ -5,6 +5,25 @@ use aptos_metrics_core::{register_int_counter_vec, IntCounterVec};
 use once_cell::sync::Lazy;
 
 /// Counter for successful telemetry events
+pub(crate) static APTOS_TELEMETRY_SERVICE_SUCCESS: Lazy<IntCounterVec> = Lazy::new(|| {
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/metrics.rs:8]: What's the difference between this and `APTOS_TELEMETRY_SUCCESS` ?

File: crates/aptos-telemetry/src/sender.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+use aptos_logger::{debug, error};
+use aptos_telemetry_service::types::{
+    auth::{AuthRequest, AuthResponse},
+    telemetry::TelemetryDump,
+};
+use aptos_types::{chain_id::ChainId, PeerId};
+use reqwest::{StatusCode, Url};
+use std::sync::Arc;
+
+// const METRICS_ENDPOINT: &str = "https://aptos.dev/telemetry/";
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:21]: Please remove

File: crates/aptos-telemetry/src/sender.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    token: RwLock<Option<String>>,
+    server_public_key: Mutex<Option<x25519::PublicKey>>,
+}
+
+impl AuthContext {
+    fn new(node_config: &NodeConfig) -> Self {
+        Self {
+            noise_config: node_config.identity_key().map(NoiseConfig::new),
+            token: RwLock::new(None),
+            server_public_key: Mutex::new(None),
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:41]: Why is this `None`? Is it possible to get the `server_public_key` in here instead?

File: crates/aptos-telemetry/src/sender.rs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        Self {
+            chain_id,
+            peer_id: node_config.peer_id().unwrap_or(PeerId::ZERO),
+            client: reqwest::Client::new(),
+            auth_context: Arc::new(AuthContext::new(node_config)),
+        }
+    }
+
+    pub async fn send_metrics(&self, event_name: String, telemetry_dump: TelemetryDump) {
+        for i in 1..MAX_RETRIES {
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:65]: Better to use a retry policy here with some exponential backoff - https://github.com/aptos-labs/aptos-core/blob/e17fa8823787cd81fba22aebaf30f71ff4466411/crates/transaction-emitter-lib/src/emitter/mod.rs#L60

File: crates/aptos-telemetry/src/sender.rs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    noise_config: Option<NoiseConfig>,
+    token: RwLock<Option<String>>,
+    server_public_key: Mutex<Option<x25519::PublicKey>>,
+}
+
+impl AuthContext {
+    fn new(node_config: &NodeConfig) -> Self {
+        Self {
+            noise_config: node_config.identity_key().map(NoiseConfig::new),
+            token: RwLock::new(None),
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:40]: Same for the token, is it possible to initialize this to a valid token instead of None?

File: crates/aptos-telemetry/src/sender.rs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                    Err(anyhow!("Error status received: {}", status_code))
+                }
+            }
+            Err(error) => Err(anyhow!("Error sending metrics. Err: {}", error)),
+        }
+    }
+
+    async fn get_token(&self) -> Result<String, Error> {
+        // Try to read the token holding a read lock
+        let token = { self.auth_context.token.read().as_ref().cloned() };
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:119]: nit: unnecessary curly braces

File: crates/aptos-telemetry/src/sender.rs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+            None => {
+                let public_key = self.get_public_key_from_server().await?;
+                *self.auth_context.server_public_key.lock() = Some(public_key);
+                Ok(public_key)
+            }
+        }
+    }
+
+    fn reset_token(&self) {
+        *self.auth_context.token.write() = None
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:156]: Instead of resetting this to `None`, should you generate a new valid token instead?

File: crates/aptos-telemetry/src/sender.rs [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+#[derive(Clone)]
+pub(crate) struct TelemetrySender {
+    chain_id: ChainId,
+    peer_id: PeerId,
+    client: reqwest::Client,
+    auth_context: Arc<AuthContext>,
+}
+
+impl TelemetrySender {
```
Comment (sitalkedia) [inline crates/aptos-telemetry/src/sender.rs:54]: Possible to add some unit tests for the sender?

File: config/src/config/mod.rs [hunk 9]
```diff
@@ -279,6 +279,20 @@ impl NodeConfig {
         }
     }
 
+    pub fn identity_key(&self) -> Option<x25519::PrivateKey> {
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply config/src/config/mod.rs:282]: I am not quite sure if we should fail in this case. Both `validator_network` and `full_node_networks` are `Option` types so it made sense to return an option as well. Also, for full node network, I am not sure if we will have a public network id configured always, so it has to be option in that case at least.

File: crates/aptos-telemetry/src/metrics.rs (hunk 1 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/metrics.rs:8]: The naming might not be the best, but since I am keeping the old logic to post metrics to Google Analytics, I am creating new counters for the telemetry service instead of using the existing ones.

File: crates/aptos-telemetry/src/sender.rs (hunk 5 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs:40]: It is possible to initialize this token here, but I would need to make a network call. I dont really like making network calls during object construction.

File: crates/aptos-telemetry/src/sender.rs (hunk 3 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs:41]: Same as with token, I would need to make a network call.

File: crates/aptos-telemetry/src/sender.rs (hunk 7 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs:156]: It might make the control flow hard to follow. Right now we have a flow like: get_token() -> authenticate() -> get_public_key() -> post metrics -> reset_token(). If I refresh the token here, then we will have two place where authentication happens - once in get_token and once in reset_token - and we will need to handle errors for reset_token as well. We anyhow wrap all this with a retry mechanism, so when there is a retry, the token will be refreshed.

File: crates/aptos-telemetry/src/sender.rs (hunk 8 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs:48]: Yes, added some.

File: crates/aptos-telemetry/src/sender.rs (hunk 4 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs]: Yes, using tokio-retry because of async'ness.

File: crates/aptos-telemetry-service/src/custom_telemetry.rs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            context.gcp_bq_config.table_id.as_str(),
+            insert_request,
+        )
+        .await
+        .map_err(|e| {
+            error!("Error due to {}", e);
+            reject::reject()
+        })?;
+
+    info!("inject succeeded {:?}", row.to_string());
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/aptos-telemetry-service/src/custom_telemetry.rs]: Changed to debug!

File: crates/aptos-telemetry-service/src/custom_telemetry.rs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    claims: Claims,
+    body: TelemetryDump,
+) -> anyhow::Result<impl Reply, Rejection> {
+    if body.user_id != claims.peer_id.to_string() {
+        return Err(reject::reject());
+    }
+
+    let mut insert_request = TableDataInsertAllRequest::new();
+
+    let telemetry_event = body.events[0].clone();
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/aptos-telemetry-service/src/custom_telemetry.rs]: I am now borrowing.

File: crates/aptos-telemetry-service/src/context.rs [hunk 12]
```diff
@@ -13,17 +14,27 @@ pub struct Context {
...
+    pub gcp_bq_config: GCPBigQueryConfig,
+
     pub jwt_encoding_key: EncodingKey,
     pub jwt_decoding_key: DecodingKey,
 }
 
 impl Context {
-    pub fn new(config: &AptosTelemetryServiceConfig, validator_cache: ValidatorSetCache) -> Self {
+    pub fn new(
+        config: &AptosTelemetryServiceConfig,
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply crates/aptos-telemetry-service/src/context.rs]: Sounds good.

File: crates/aptos-telemetry/src/sender.rs (hunk 6 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs:126]: Are you sure the read guard would get dropped? I am getting a write lock below so I need to make sure this read lock is dropped.

File: crates/aptos-telemetry/src/metrics.rs (hunk 1 above)
Comment (sitalkedia) [reply crates/aptos-telemetry/src/metrics.rs:8]: Let's add a comment here for clarity and also I am assuming the old counters will be deprecated soon, can you add a TODO as well?

File: crates/aptos-telemetry/src/sender.rs (hunk 5 above)
Comment (sitalkedia) [reply crates/aptos-telemetry/src/sender.rs:33]: I think there is no issue in making a network call in a constructor - it is no different than any ordinary function call.  I am okay with this for now, but it would be good to clean up and simplify the logic of the sender a bit as a follow up.

**Review by sitalkedia — APPROVED** [review]
@ibalajiarun - As discussed offline, stamping this to unblock you for release today. As a follow up item, it would be good to clean up and simplify the logic in `sender.rs`

File: crates/aptos-telemetry/src/sender.rs (hunk 5 above)
Comment (ibalajiarun) [reply crates/aptos-telemetry/src/sender.rs:33]: sounds good!
